swift/ql/lib/codeql/swift/elements/decl/EnumDecl.qll

```diff
@@ -1,4 +1,51 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.decl.EnumDecl
+private import codeql.swift.elements.decl.EnumCaseDecl
+private import codeql.swift.elements.decl.EnumElementDecl
+private import codeql.swift.elements.decl.Decl
 
-class EnumDecl extends Generated::EnumDecl { }
+/**
+ * An enumeration declaration, for example:
+ * ```
+ * enum MyColours {
+ *   case red
+ *   case green
+ *   case blue
+ * }
+ * ```
+ */
+class EnumDecl extends Generated::EnumDecl {
+  /**
+   * Gets the number of `EnumElementDecl`s in this enumeration before the `memberIndex`th member. Some
+   * of the members of an `EnumDecl` are `EnumCaseDecls` (representing the `case` lines), each of
+   * which holds one or more `EnumElementDecl`s.
+   */
+  private int countEnumElementsTo(int memberIndex) {
+    memberIndex = 0 and result = 0
+    or
+    exists(Decl prev | prev = this.getMember(memberIndex - 1) |
+      result = this.countEnumElementsTo(memberIndex - 1) + prev.(EnumCaseDecl).getNumberOfElements()
+      or
+      not prev instanceof EnumCaseDecl and
+      result = this.countEnumElementsTo(memberIndex - 1)
+    )
+  }
+
+  /**
+   * Gets the `index`th enumeration element of this enumeration (0-based).
+   */
+  final EnumElementDecl getEnumElement(int index) {
+    exists(int memberIndex |
+      result =
+        this.getMember(memberIndex)
+            .(EnumCaseDecl)
+            .getElement(index - this.countEnumElementsTo(memberIndex))
+    )
+  }
```

Comment:
This seems to be a pretty perfect case for a `rank` aggregate. It's shorter and (to me, at least) more clear, and likely performs much better than the manual recursive version.
```suggestion
  /**
   * Gets the `index`th enumeration element of this enumeration (0-based).
   */
  final EnumElementDecl getEnumElement(int index) {
    result =
      rank[index + 1](int memberIndex, Decl d |
        d = this.getMember(memberIndex) and
        d instanceof EnumElementDecl
      |
        d order by memberIndex
      )
  }
```

Reply:
Your code looks good but seems a bit _too_ simple ... I think it's missing the `getElement` (`EnumCaseDecl` -> `EnumElementDecl`) step.  I'll investigate tomorrow.  I'd definitely prefer something like this in principle to my initial attempt.

Reply:
Oh I see how it works now ... I'd completely missed that the `EnumElementDecl`s are direct children of the `EnumDecl`, not just of the `EnumCaseDecl`.  Your fix LGTM (and passes the tests).